### PR TITLE
WIP.

### PR DESCRIPTION
**Pull request by chris-wood**

### CODE REVIEW

File: draft-ietf-tls-esni.md [hunk 1]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
 and either terminate the connection (in Shared Mode) or forward
 it to the backend server (in Split Mode).
 
-# Publishing the SNI Encryption Key {#publishing-key}
+# Publishing the SNI Encryption Key in DNS {#publishing-key}
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md]: nit: in the DNS

File: draft-ietf-tls-esni.md [hunk 2]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
 
-SNI Encryption keys can be published in the DNS using the ESNIKeys
-structure, defined below.
+Publishing ESNI keys in DNS requires care to ensure correct behavior.
+At least two requirements must be met. First, resolution of an ESNIKeys
+record, described in {{esni-record}}, must yield information necessary for
+deterministic client connections to a host which is guaranteed to possess
+the corresponding ESNI private key. This is necessary to avoid ESNI failure
+fallback. Servers MUST ensure that if multiple A or AAAA records are returned
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md]: I don't think you've defined what fallback means at this point in the draft

Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md]: I'm confused, is this paragraph describing the rationale for the solution, or the mechanism itself? If the former, then there probably shouldn't be normative text here

Comment (ekr) [inline draft-ietf-tls-esni.md]: What's failure fallback?

File: draft-ietf-tls-esni.md [hunk 3]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+for a domain with ESNI support, all the servers pointed to by those records are
+able to handle the keys returned as part of a ESNIKeys record for that domain.
+
+Servers operating in Split Mode SHOULD have DNS configured to return
+the same A (or AAAA) record for all ESNI-enabled servers they service. This yields
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md:190]: I'm not sure this work with DNS-based load balancing, where there is a long list of servers and the subset you get in the DNS depends on load or randomness. I think the property you want is: the addresses returned by the DNS (in A and AAAA records) for ESNI-enabled TLS servers SHOULD not depend on the queried domain.

File: draft-ietf-tls-esni.md [hunk 4]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+an anonymity set of cardinality equal to the number of ESNI-enabled server domains
+supported by a given client-facing server. Thus, even with SNI encryption,
+an attacker which can enumerate the set of ESNI-enabled domains supported
+by a client-facing server can guess the correct SNI with probability at least
+1/K, where K is the size of this ESNI-enabled server anonymity set. This probability
+may be increased via traffic analysis or other mechanisms.
+
+Secondly, and relatedly, presence of DNS-based load balancers
+such as Cedexis, which distribute DNS queries across different CDN providers,
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md]: I wouldn't name vendors in a draft this way

File: draft-ietf-tls-esni.md [hunk 5]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+must not lead to ESNI key and A/AAAA record provider mismatches. That is,
+it must not be possible for clients to receive A/AAAA records that point to
+one service provider, while receiving ESNI keys that refer to a different
+provider.
+
+The following sections describe a DNS record format that achieve these goals.
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md:198]: Now that I've finished this section, I feel like it was discussing rationale - but in that case I'd call that out in the section title and remove all normative text

File: draft-ietf-tls-esni.md [hunk 6]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
-SNI Encryption keys can be published in the DNS using the ESNIKeys
-structure, defined below.
+Publishing ESNI keys in DNS requires care to ensure correct behavior.
+At least two requirements must be met. First, resolution of an ESNIKeys
+record, described in {{esni-record}}, must yield information necessary for
+deterministic client connections to a host which is guaranteed to possess
+the corresponding ESNI private key. This is necessary to avoid ESNI failure
+fallback. Servers MUST ensure that if multiple A or AAAA records are returned
+for a domain with ESNI support, all the servers pointed to by those records are
+able to handle the keys returned as part of a ESNIKeys record for that domain.
```
Comment (ekr) [inline draft-ietf-tls-esni.md]: This doesn't seem like quite the right requirement. I think I would try to set the problem up first.

"There are deployment environments in which a domain is served by multiple server operators who do not sure the ESNI Keys. Because ESNIKeys and A/AAAA lookup are independent, it is therefore possible to obtain an ESNIKeys record which does not match the A/AAAA records. Naively using this record would result in handshake failure. The design of the system must therefore allow clients to detect and recover from this situation."

File: draft-ietf-tls-esni.md [hunk 7]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+
+Servers operating in Split Mode SHOULD have DNS configured to return
+the same A (or AAAA) record for all ESNI-enabled servers they service. This yields
+an anonymity set of cardinality equal to the number of ESNI-enabled server domains
+supported by a given client-facing server. Thus, even with SNI encryption,
+an attacker which can enumerate the set of ESNI-enabled domains supported
+by a client-facing server can guess the correct SNI with probability at least
+1/K, where K is the size of this ESNI-enabled server anonymity set. This probability
+may be increased via traffic analysis or other mechanisms.
```
Comment (ekr) [inline draft-ietf-tls-esni.md:196]: This seems like good advice, but why is it just for split mode? And why does it belong here?

File: draft-ietf-tls-esni.md [hunk 8]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+
+Secondly, and relatedly, presence of DNS-based load balancers
+such as Cedexis, which distribute DNS queries across different CDN providers,
+must not lead to ESNI key and A/AAAA record provider mismatches. That is,
+it must not be possible for clients to receive A/AAAA records that point to
+one service provider, while receiving ESNI keys that refer to a different
+provider.
```
Comment (ekr) [inline draft-ietf-tls-esni.md]: I think I actually covered this above.

File: draft-ietf-tls-esni.md [hunk 9]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+
+The following sections describe a DNS record format that achieve these goals.
+
+## Host Pointers
+
+Encrypted SNI records, described in {{esni-record}}, point to the (set of) hosts which possess
+the private ESNI key. These pointers include (1) a canonical name which, if resolved, MUST yield A
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md:203]: Is "canonical name" the same as davidben's "public name"? If so I liked "public name" better

File: draft-ietf-tls-esni.md [hunk 10]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+    struct {
+        AddressType address_type;
+        select (address_type) {
+            case address_v4: {
+                opaque ipv4Address[4];
+                opaque ipv4Mask[4];
+            }
+            case address_v6: {
+                opaque ipv6Address[16];
+                opaque ipv6Mask[16];
```
Comment (ekr) [inline draft-ietf-tls-esni.md]: These masks should be a length.

File: draft-ietf-tls-esni.md [hunk 11]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+
+~~~~
+    enum {
+        address_v4,
+        address_v6,
+    } AddressType;
+
+    struct {
+        AddressType address_type;
+        select (address_type) {
```
Comment (ekr) [inline draft-ietf-tls-esni.md:217]: Cool people have a length field here so you can skip unknown types.

File: draft-ietf-tls-esni.md [hunk 12]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+
+    struct {
+        HostName name;
+        Address address_set<0..2^16-1>;
+    } HostPointer;
+~~~~
+
+name
+: The terminal name for which resolution MUST yield valid A or
+AAAA records pointing to a host which holds the private ESNI key.
```
Comment (ekr) [inline draft-ietf-tls-esni.md:237]: This seems like it needs to be unpacked. Like, why is this OK.

File: draft-ietf-tls-esni.md [hunk 13]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+
+address_set
+: An optional list of Address entries, each containing a single IPv4 or IPv6
+SIDR block, possibly of length 1.
```
Comment (ekr) [inline draft-ietf-tls-esni.md]: ITYM CIDR.

File: draft-ietf-tls-esni.md [hunk 14]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+or AAAA records that possess the corresponding private key, and (2) a set of CIDR blocks,
+each possibly of size 1, that match the IPv4 or IPv6 address of a host with the private key.
+
+These "host pointers" are encoded using the following structure.
+
+~~~~
+    enum {
+        address_v4,
+        address_v6,
+    } AddressType;
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md:213]: I'd just use the IP version - this enum takes a full byte anyway

File: draft-ietf-tls-esni.md [hunk 15]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+    } Address;
+
+    struct {
+        HostName name;
+        Address address_set<0..2^16-1>;
+    } HostPointer;
+~~~~
+
+name
+: The terminal name for which resolution MUST yield valid A or
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md]: I don't think "terminal name" is standard DNS terminology, so you'll need to define it in this draft

Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md]: I don't think you can put a MUST here. Resolution is a process involving many different bits of infrastructure. RFC 2119 normative language only applied to implementations of this document

File: draft-ietf-tls-esni.md [hunk 16]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+        address_v4,
+        address_v6,
+    } AddressType;
+
+    struct {
+        AddressType address_type;
+        select (address_type) {
+            case address_v4: {
+                opaque ipv4Address[4];
+                opaque ipv4Mask[4];
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md]: The text above this diagram mentions CIDR, the text below mentions SIDR (OK that's a typo) but this isn't CIDR. CIDR would be an address followed by a prefix length. I would recommend using either a 1-byte prefix length, or using address ranges instead, as in IKEv2 traffic selectors https://tools.ietf.org/html/rfc5996#section-3.13.1

File: draft-ietf-tls-esni.md [hunk 17]
```diff
@@ -302,6 +359,41 @@ servers to rotate the keys often and improve forward secrecy.
 Note that the length of this structure MUST NOT exceed 2^16 - 1, as the
 RDLENGTH is only 16 bits {{RFC1035}}.
 
+## Encrypted SNI DNS Resolution {#esni-resolution}
+
+Clients obtain ESNI records by querying DNS for ESNI-enabled server domains.
+In cases where the domain of the A or AAAA records being resolved do not match the
+SNI Server Name, such as when {{!RFC7838}} is being used, the SNI domain should
+be used for querying the ESNI TXT record.
```
Comment (ekr) [inline draft-ietf-tls-esni.md:361]: Oh, this is an interesting catch. Say that I go to ```www.example.com``` and I get ```alt.example.com```. You're saying that I lookup:

```
alt.example.com? A/AAAA
www.example.com? ESNIKeys
```

So, I'm actually not sure that this is the right answer, as it depends on how alt-svc is being used. Consider the case where the origin is pointing to a CDN. Don't we want the CDN's ESNIKeys to control?

In either case, you should use "origin" and "alternative"

File: draft-ietf-tls-esni.md [hunk 18]
```diff
@@ -174,10 +174,83 @@ provider's public key. The provider can then decrypt the extension
...
+                opaque ipv6Address[16];
+                opaque ipv6Mask[16];
+            }
+        }
+    } Address;
+
+    struct {
+        HostName name;
+        Address address_set<0..2^16-1>;
+    } HostPointer;
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md:232]: Pointer might get confused with DNS PTR records, I'd use a different name

File: draft-ietf-tls-esni.md [hunk 19]
```diff
@@ -302,6 +359,41 @@ servers to rotate the keys often and improve forward secrecy.
...
 RDLENGTH is only 16 bits {{RFC1035}}.
 
+## Encrypted SNI DNS Resolution {#esni-resolution}
+
+Clients obtain ESNI records by querying DNS for ESNI-enabled server domains.
+In cases where the domain of the A or AAAA records being resolved do not match the
+SNI Server Name, such as when {{!RFC7838}} is being used, the SNI domain should
+be used for querying the ESNI TXT record.
+
+Clients SHOULD initiate these queries in parallel alongside normal A or AAAA queries,
```
Comment (ekr) [inline draft-ietf-tls-esni.md]: these queries -> ESNI queries.

File: draft-ietf-tls-esni.md [hunk 20]
```diff
@@ -302,6 +359,41 @@ servers to rotate the keys often and improve forward secrecy.
...
+preferring AAAA if possible. Specifically, clients should order their queries as
+AAAA, ESNI, and A, thereby preferring IPv6 and ESNI over IPv4. Clients SHOULD block
```
Comment (ekr) [inline draft-ietf-tls-esni.md]: Let's not get into an IPv6/IPv4 debate here.

File: draft-ietf-tls-esni.md [hunk 21]
```diff
@@ -302,6 +359,41 @@ servers to rotate the keys often and improve forward secrecy.
...
+TLS handshakes until complete perhaps by timing out. If responses arrive
```
Comment (ekr) [inline draft-ietf-tls-esni.md]: I think you need a more concrete algorithm here, and to be clearer about what you want.

File: draft-ietf-tls-esni.md [hunk 22]
```diff
@@ -302,6 +359,41 @@ servers to rotate the keys often and improve forward secrecy.
...
+use full addresses, said client must compare ESNIKeys.host_pointer.name to the result of the A and AAAA queries.
+2a. If neither the A or AAAA queries yielded a CNAME, then ESNIKeys.host_pointer.name MUST match
+the server domain name used in the A and AAAA query. If it does not, return an error.
+2b. If an A or AAAA query yielded a CNAME, and ESNIKeys.host_pointer.name matches this CNAME,
+the client MUST resolve this CNAME to an A or AAAA answer, and return the resulting addresses.
+2c. If an A or AAAA query yielded a CNAME, and ESNIKeys.host_pointer.name does not match this CNAME,
+and the netmask in the corresponding ESNIKeys.host_pointer address matches the A or AAAA result
+from resolving this CNAME, return these addresses.
+2d. Return an error, as the ESNIKeys record does not contain an address or address mask that matches
+the corresponding A or AAAA query results.
```
Comment (ekr) [inline draft-ietf-tls-esni.md]: I'm not sure that this algorithm is quite what we want. 

Stepping back, I think you want something like this:

1. If you get ESNIKeys and it's got a full address, use the address.
2. If you get A or AAAA, wait up to X seconds for ESNIKeys and then if you don't get it, proceed without. If you do get it, and it's got a full address, use that, as in (2).
3. If you have ESNIKeys and either A or AAAA and the address mask matches, then use that address and ESNIKeys, possibly waiting for up to Y seconds if you have only A or AAAA and prefer the other type of address.
4. If you have ESNIKeys and either A or AAAA and the address mask doesn't match then issue a query for the Hostname and that type of address.
5. If you do all the queries mentioned above and you still don't have an address match, then error.

I don't think you need the CNAME comparisons. How do they help?

**Review by ekr — COMMENTED** [review]
A bunch of comments here. At a high level, I'm not sure the union is the right approach. I'd like to see an estimate of the actual size of an ESNIKeys record for a typical CDN.

File: draft-ietf-tls-esni.md (hunk 2 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: Fallback to plaintext SNI.

File: draft-ietf-tls-esni.md (hunk 6 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: Text LGTM! Will adopt.

File: draft-ietf-tls-esni.md (hunk 7 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:196]: Oh, this is just copy-pasted. We should revisit this in a separate issue?

File: draft-ietf-tls-esni.md (hunk 8 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: 👍

File: draft-ietf-tls-esni.md (hunk 13 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: Yep.

File: draft-ietf-tls-esni.md (hunk 19 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: 👍

File: draft-ietf-tls-esni.md (hunk 20 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: I'll remove it.

File: draft-ietf-tls-esni.md [hunk 23]
```diff
@@ -302,6 +359,41 @@ servers to rotate the keys often and improve forward secrecy.
 Note that the length of this structure MUST NOT exceed 2^16 - 1, as the
 RDLENGTH is only 16 bits {{RFC1035}}.
 
+## Encrypted SNI DNS Resolution {#esni-resolution}
+
+Clients obtain ESNI records by querying DNS for ESNI-enabled server domains.
+In cases where the domain of the A or AAAA records being resolved do not match the
+SNI Server Name, such as when {{!RFC7838}} is being used, the SNI domain should
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md]: I don't think Alt-Svc means you put the origin in the SNI when handshaking with the alternate server.
https://tools.ietf.org/html/rfc7838#section-2.1

File: draft-ietf-tls-esni.md (hunk 19 above)
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md]: This violates many many layers and is going to be incredibly hard to implement, as many vendors implement those systems in separate codebases. I'm worried this will get significant pushback at the IETF

File: draft-ietf-tls-esni.md (hunk 2 above)
Comment (tfpauly) [reply draft-ietf-tls-esni.md]: `Failure fallback` does seem hard to parse. Perhaps just write out `This is necessary to avoid falling back from ESNI to plaintext SNI.`

File: draft-ietf-tls-esni.md (hunk 17 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:361]: I think we should address this in a separate issue as it's unrelated to this change. (Migrated text.)

File: draft-ietf-tls-esni.md (hunk 22 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: The CNAME comparison is for the case where the address set is empty or the mask does not match, yet the A/AAAA yielded a terminal CNAME that matches the contents of the ESNIKeys structure. If there's a match there, then clients may connect to the address, even though there's a netmask mismatch. I *think* this works.

File: draft-ietf-tls-esni.md (hunk 2 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: 👍 Fixed.

File: draft-ietf-tls-esni.md (hunk 3 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:190]: This particular bit is sort of unrelated to these load balancers. I tried to tidy it up in the latest push.

File: draft-ietf-tls-esni.md (hunk 15 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: Fine, fine. :)

File: draft-ietf-tls-esni.md (hunk 16 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: I'm just going with the 1 byte prefix length.

File: draft-ietf-tls-esni.md (hunk 18 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:232]: Recommendations?

File: draft-ietf-tls-esni.md (hunk 23 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: Copied text. Please file an open issue against it!

Comment (chris-wood) [reply draft-ietf-tls-esni.md]: (Also, I think we cleared this up yesterday?)

File: draft-ietf-tls-esni.md [hunk 24]
```diff
@@ -302,6 +353,58 @@ servers to rotate the keys often and improve forward secrecy.
...
+## Encrypted SNI DNS Resolution {#esni-resolution}
+
+Clients obtain ESNI records by querying DNS for ESNI-enabled server domains.
+In cases where the domain of the A or AAAA records being resolved do not match the
+SNI Server Name, such as when {{!RFC7838}} is being used, the alternate domain should
+be used for querying the ESNI TXT record.
+
+Clients SHOULD initiate ESNI queries in parallel alongside normal A or AAAA queries.
+There are generally two cases that determine how clients should handle ESNIKeys
+responses: (1) ESNIKeys have full addresses, and (2) ESNIKeys have address net masks.
```
Comment (tfpauly) [inline draft-ietf-tls-esni.md]: Nit rather than `have`, say `contain`?

File: draft-ietf-tls-esni.md [hunk 25]
```diff
@@ -302,6 +353,58 @@ servers to rotate the keys often and improve forward secrecy.
...
+The following algorithm describes how clients should process ESNIKeys responses with full
+addresses as they arrive.
+
+~~~
+1. If an ESNIKeys response arrives before an A or AAAA response, clients should connect to said address.
+2. If an A or AAAA response arrives before the ESNIKeys response, clients should wait up
+to CD seconds before attempting to initiate a connection to an address. If an ESNIKeys response
+does not arrive in this time, clients should initiate a connection to the provided address(es).
+If an ESNIKeys response does arrive in this time, clients should connect to said address.
```
Comment (tfpauly) [inline draft-ietf-tls-esni.md]: Should you explain that if the ESNIKeys arrives after the TCP connection is attempted, but before the TLS handshake begins, and it matches the addresses, you should still use ESNI?

File: draft-ietf-tls-esni.md [hunk 26]
```diff
@@ -302,6 +353,58 @@ servers to rotate the keys often and improve forward secrecy.
...
+1. If an A or AAAA response arrives before the ESNIKeys response, clients should wait up
+to CD seconds before attempting to initiate a connection to an address. If an ESNIKeys response
+does not arrive in this time, clients should initiate a connection to the provided address(es).
+2. If an ESNIKeys response arrives before this time, and if the address netmask matches that
+in the corresponding A or AAAA response, then clients should connect to said address.
+Clients may delay this check until both A and AAAA responses arrive, or if a Resolution Delay
+timeout occurs, according to {{RFC8305}}.
+3. If the address netmask does not match the corresponding A or AAAA response, and the
+A or AAAA response yielded a terminal CNAME that matches ESNIKeys.host_pointer.name,
+then clients should connect to said address.
```
Comment (tfpauly) [inline draft-ietf-tls-esni.md]: `said address` is getting confusing here now that there are multiple addresses (A/AAAA and ESNIKeys) floating around

File: draft-ietf-tls-esni.md (hunk 23 above)
Comment (DavidSchinazi) [reply draft-ietf-tls-esni.md]: #132

File: draft-ietf-tls-esni.md (hunk 17 above)
Comment (DavidSchinazi) [reply draft-ietf-tls-esni.md:361]: #132

File: draft-ietf-tls-esni.md [hunk 27]
```diff
@@ -302,6 +353,57 @@ servers to rotate the keys often and improve forward secrecy.
...
+
+3. If an ESNIKeys response arrives before this time, and if the address netmask
+matches that in the corresponding A or AAAA response, then clients should initiate
+TLS with ESNI to said address(es). (An ESNIKeys record with a full address has a
+  full netmask, and therefore the match must be exact.) Clients may delay this
+  check until both A and AAAA responses arrive, or if a Resolution Delay timeout
+  occurs, according to {{RFC8305}}.
+
+4. If the address netmask does not match the corresponding A or AAAA response,
+and the A or AAAA response yielded a terminal CNAME that matches ESNIKeys.host_pointer.name,
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md:386]: This is the part that could be a serious issue for implementors. When you use `getaddrinfo()` you don't have access to CNAMEs that could have occurred. I would make this step optional

File: draft-ietf-tls-esni.md [hunk 28]
```diff
@@ -302,6 +353,57 @@ servers to rotate the keys often and improve forward secrecy.
...
+then clients should initiate TLS using ESNI to the address(es) in the A or AAAA records.
+
+5. If the address netmask does not match the corresponding A or AAAA response,
+and neither the A or AAAA queries yielded a CNAME that matches ESNIKeys.host_pointer.name,
+clients should resolve the address of ESNIKeys.host_pointer.name using an
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md:391]: This sounds like slow and complicated machinery that is unnecessary - why don't you just require the ESNI record to contain full addresses?

File: draft-ietf-tls-esni.md [hunk 29]
```diff
@@ -302,6 +353,57 @@ servers to rotate the keys often and improve forward secrecy.
...
+using ENSI to the resulting address(es) when A or AAAA resolution completes.
+
+6. If the ESNIKeys.host_pointer address set is empty, clients should resolve the
+address(es) of ESNIKeys.host_pointer.name and initiate TLS to the resulting address(es).
+
+7. In all other cases, raise an error.
+~~~
+
+CD (Connection Delay) is a configurable parameter. The recommended value is 50 milliseconds,
+as per the guidance in {{!RFC8305}}. This algorithm ensures that clients can deterministically
```
Comment (DavidSchinazi) [inline draft-ietf-tls-esni.md]: This isn't deterministic, on the contrary it's probabilistic - the only value of the 50ms delay in HEv2 is to move the probability of using IPv6 from 50% to 99%. You'd get similar results here

File: draft-ietf-tls-esni.md (hunk 24 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: Fixed!

File: draft-ietf-tls-esni.md (hunk 25 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: Fixed!

File: draft-ietf-tls-esni.md (hunk 28 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:391]: As discussed, not every operator can do that.

File: draft-ietf-tls-esni.md (hunk 29 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: Deterministic isn't the right word -- I mean to say it converges.

File: draft-ietf-tls-esni.md (hunk 27 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:386]: Indeed, it can be, per client policy. I'll rewrite the "should" bit for the algorithm to say that anything beyond use of full addresses is not needed.
